multicluster/cmd/multicluster-controller/memberclusterannounce_webhook.go

```diff
@@ -54,33 +56,57 @@ func (v *memberClusterAnnounceValidator) Handle(ctx context.Context, req admissi
 		return admission.Errored(http.StatusBadRequest, err)
 	}
 
-	// read the ClusterSet info
-	clusterSetList := &multiclusterv1alpha1.ClusterSetList{}
-	if err := v.Client.List(context.TODO(), clusterSetList, client.InNamespace(v.namespace)); err != nil {
-		klog.ErrorS(err, "Error reading ClusterSet", "Namespace", v.namespace)
+	serviceAccount := &v1.ServiceAccount{}
+	if err := v.Client.Get(ctx, client.ObjectKey{Namespace: v.namespace, Name: saName}, serviceAccount); err != nil {
+		klog.ErrorS(err, "Error getting ServiceAccount", "ServiceAccount", saName, "Namespace", v.namespace, "MemberClusterAnnounce", klog.KObj(memberClusterAnnounce))
 		return admission.Errored(http.StatusPreconditionFailed, err)
 	}
 
-	if len(clusterSetList.Items) != 1 {
-		return admission.Errored(http.StatusPreconditionFailed,
-			fmt.Errorf("invalid ClusterSet config in the leader cluster, please contact your administrator"))
+	var newObj *multiclusterv1alpha1.MemberClusterAnnounce
+	var oldObj *multiclusterv1alpha1.MemberClusterAnnounce
+	if req.Object.Raw != nil {
+		if err := json.Unmarshal(req.Object.Raw, &newObj); err != nil {
+			klog.ErrorS(err, "Error while decoding new MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(memberClusterAnnounce))
+			return admission.Errored(http.StatusBadRequest, err)
+		}
 	}
-
-	clusterSet := clusterSetList.Items[0]
-	if clusterSet.Name == memberClusterAnnounce.ClusterSetID {
-		for _, member := range clusterSet.Spec.Members {
-			if member.ClusterID == memberClusterAnnounce.ClusterID {
-				// validate the ServiceAccount used is correct
-				if member.ServiceAccount == saName {
-					return admission.Allowed("")
-				} else {
-					return admission.Denied("Member does not have permissions")
-				}
-			}
+	if req.OldObject.Raw != nil {
+		if err := json.Unmarshal(req.OldObject.Raw, &oldObj); err != nil {
+			klog.ErrorS(err, "Error while decoding old MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(memberClusterAnnounce))
+			return admission.Errored(http.StatusBadRequest, err)
 		}
 	}
 
-	return admission.Denied("Unknown member")
+	switch req.Operation {
+	case admissionv1.Create:
+		// read the ClusterSet info
+		clusterSetList := &multiclusterv1alpha1.ClusterSetList{}
+		if err := v.Client.List(context.TODO(), clusterSetList, client.InNamespace(v.namespace)); err != nil {
+			klog.ErrorS(err, "Error reading ClusterSet", "Namespace", v.namespace)
+			return admission.Errored(http.StatusPreconditionFailed, err)
+		}
+
+		clusterSet := clusterSetList.Items[0]
+		if clusterSet.Name == memberClusterAnnounce.ClusterSetID &&
+			clusterSet.Spec.Leaders[0].ClusterID == memberClusterAnnounce.LeaderClusterID {
+			return admission.Allowed("")
+		}
+
+		if clusterSet.Name != memberClusterAnnounce.ClusterSetID {
+			return admission.Denied("Unknown ClusterSet ID")
+		}
+		if clusterSet.Spec.Leaders[0].ClusterID != memberClusterAnnounce.LeaderClusterID {
```

Comment:
My earlier question was removed by code update. Repaste it:

> @luolanzone : why we need to include leader cluster ID in the MemberClusterAnnounce? Could we remove it?

Reply:
I double checked the code, it's used for MemberClusterAnnounce controller to validate and update the status https://github.com/antrea-io/antrea/blob/9f76f8eb303f6147353b2b62200418aef75b2cb3/multicluster/controllers/multicluster/memberclusterannounce_controller.go#L115.
I feel it can be removed if we allow member to be added into leader ClusterSet without checking ServiceAccount.
I need double check this part. We probably can have another PR to remove it if we confirm it's Ok to be removed